src/legacy/core_plugins/kibana/public/visualize/np_ready/legacy_app.js

```diff
@@ -144,19 +178,57 @@ export function initVisualizeApp(app, deps) {
         template: editorTemplate,
         k7Breadcrumbs: getEditBreadcrumbs,
         resolve: {
-          savedVis: function($route, $rootScope, kbnUrl, history) {
-            const { chrome, core, data, savedVisualizations, toastNotifications } = deps;
+          resolved: function($route, $rootScope, kbnUrl, history) {
+            const {
+              chrome,
+              core,
+              data,
+              savedVisualizations,
+              visualizations,
+              toastNotifications,
+            } = deps;
+
+            const results = {};
+
             return ensureDefaultIndexPattern(core, data, $rootScope, kbnUrl)
               .then(() => savedVisualizations.get($route.current.params.id))
               .then(savedVis => {
                 chrome.recentlyAccessed.add(savedVis.getFullPath(), savedVis.title, savedVis.id);
-                return savedVis;
+                results.savedVis = savedVis;
+                return visualizations.convertToSerializedVis(savedVis);
               })
-              .then(savedVis => {
-                if (savedVis.vis.type.setup) {
-                  return savedVis.vis.type.setup(savedVis).catch(() => savedVis);
+              .then(serializedVis => visualizations.createVis(serializedVis.type, serializedVis))
+              .then(vis => {
+                if (vis.type.setup) {
+                  return vis.type.setup(vis).catch(() => vis);
+                }
+                return vis;
+              })
+              .then(vis => {
+                results.vis = vis;
+                return deps.embeddable
+                  .getEmbeddableFactory('visualization')
+                  .createFromObject(results.vis, {
+                    timeRange: data.query.timefilter.timefilter.getTime(),
+                    filters: data.query.filterManager.getFilters(),
+                  });
+              })
+              .then(embeddableHandler => {
+                results.embeddableHandler = embeddableHandler;
+                if (results.vis.data.savedSearchId) {
+                  return createSavedSearchesLoader({
+                    savedObjectsClient: core.savedObjects.client,
+                    indexPatterns: data.indexPatterns,
+                    chrome: core.chrome,
+                    overlays: core.overlays,
+                  }).get(results.vis.data.savedSearchId);
```

Comment:
nit - Both promise chains in this file are pretty long and mostly duplicated; you might consider splitting them apart into individual functions defined in one place